### ISSUE TITLE
WSL2: interop broken in build 18945

### ISSUE THREAD

**Issue by tobiaskohlbau** [microsoft/WSL#4337]
* Your Windows build number:  Microsoft Windows [Version 10.0.18945.1001]

* What you're doing and what's happening:  Execute any windows executable from within WSL2 and redirect the output. For e.g. `cmd.exe /c "test" 2> /dev/null` or `notepad.exe 2> /dev/null`.
* What's wrong / what should be happening instead: The program return hangs forever, the executable gets called and the output is printed/ the program work but the program does not return after it's finished.

Looks like this is a regression introduced by 18945 as it worked on WSL2 before this update. If you need more information or there is something I need to provide, please let me know.

Edit: I've confirmed that only WSL2 is affected by running the same examples within a WSL1 distribution which work like expected. Also I've validated the bug within another WSL2 distribution (Pengwin and Ubuntu) and both experience the described behavior.

**Comment by simonferquel** [comment]
Related issue with same build (let me know if you prefer that I open a separate issue):

- What you're doing and what's happening: from a wsl terminal, try to run Visual Studio Code in remote to WSL mode: `code .`
- What's wrong / what should be happening instead: the terminal hangs and VS code does not show up (`ctrl+c` unblock the terminal, but vs code does not show up either). In previous build, it was my primary way to launch vs code in remote to WSL mode.

Note: launching vs code from windows and connect to WSL works fine (this is my current workaround)

**Comment by simonferquel** [comment]
I also confirm I can reproduce @tobiaskohlbau issue

**Comment by nishikawasasaki** [comment]
Same here with `10.0.18945.1001`.

When this happened, I found in `ps -ef` outputs that `/tools/init` + any `exe` stopped.
I tested below and `ctrl+c` doesn't work.
  - explorer.exe
  - code-insiders (is `Code - Insiders.exe`)
  - win32yank.exe (in tmux)

**Comment by benhillis** [comment]
I've identified the issue and am working on a fix.

**Comment by onomatopellan** [comment]
I'm glad the fix is coming. I think I have the same problem. Cross compiling with ninja does't work because it hangs at some point. Only killing the stuck .exe permits the build to continue. 

![lame](https://user-images.githubusercontent.com/599142/62840569-13d31a00-bc9d-11e9-96ae-cef6ad191c9f.png)

**Comment by pkit** [comment]
Dunno if related, but seems so: 
- run browser (chrome.exe headless) as a child process
- send SIGTERM to the init pid 
- will terminate init but leaves browser happily running.

**Comment by magneticflux-** [comment]
Still broken in 18963.

**Comment by magneticflux-** [comment]
Still broken in 18965.

**Comment by JohnButare** [comment]
For what it's worth I have the same issue with every distribution I have tried in WSL2 on build 18965 (Alpine, Ubuntu-18.04).   I can get the shell to return bt doing a ctrl-z and fg

jjbutare-wvm1:~$ ipconfig.exe | grep Ethernet
Ethernet adapter vEthernet (Default Switch):
Ethernet adapter Ethernet:
Ethernet adapter vEthernet (WSL):
^Z[1]+  Stopped                    ipconfig.exe | grep Ethernet
jjbutare-wvm1:~$ fg
ipconfig.exe | grep Ethernet
jjbutare-wvm1:~$ fg
-ash: fg: No current job
jjbutare-wvm1:~$

**Comment by thisguychris** [comment]
@benhillis sorry to ask like a 12-year-old, but what build should the fix come? Also, when did it break? I want to revert to the last known state, I have to use my other non-windows laptop to do work. :(

**Comment by JohnButare** [comment]
@benhillis - could you convert back to WSL 1?   I have not seen this issue on WSL 1, and don't personally have any must-have features in 2.   I do noticed the dramatic improvement in Linux file system access, but truthfully the Windows interop file system speed is slow - I am going to file a bug for that.   A WSL ls of system32 takes 10s in my VM, but WSL1 is like .5s.

**Comment by benhillis** [comment]
It's hard to give an estimate because there are so many factors. It will be soon though. I wish we were able to get fixes out quicker but that is the reality of a project as big as Windows.

**Comment by thisguychris** [comment]
@JohnButare I just did an ls of system32, it takes about 1.96 seconds. Currently at 18950

**Comment by Biswa96** [comment]
@thisguychris Can you provide that time in WSL1 and WSL2 both?

**Comment by thisguychris** [comment]
sorry @Biswa96 / @benhillis / @JohnButare , that screenshot I show was in WSL1, doing an ls in system32 is around 19 seconds long :(

Also, interop is broken in 18950, I guess back to wsl1 again.

**Comment by benhillis** [comment]
Fixed in [18970](https://github.com/MicrosoftDocs/WSL/releases/tag/18970).

**Comment by onomatopellan** [comment]
Thanks! I can confirm cross compiling works again in build 18970. 👌

**Comment by derkan** [comment]
Getting seg fault ar build 419104. 
Tried running notepad.exe  and code, both exited with segmentation fault

**Comment by benhillis** [comment]
@derkan - please open a new issue.
